src/System.Net.Http/src/System/Net/Http/Unix/CurlCallbacks.cs

```diff
@@ -0,0 +1,240 @@
+                    Marshal.Copy(ptr, byteBuffer, 0, byteBuffer.Length);
+                    Stream contentStream = state.ResponseMessage.ContentStream;
+                    contentStream.Write(byteBuffer, 0, byteBuffer.Length);  // will block if no readers
+                    contentStream.Flush();
+                    return totalSize;
+                });
+        }
+
+
+        private ulong CurlSendCallback(
+            IntPtr buffer,
+            ulong size,
+            ulong nitems,
+            IntPtr context)
+        {
+            return ExecuteCallback(buffer, size, nitems, context,
+                (ptr, totalSize, state) =>
+                {
+                    state.CancellationToken.ThrowIfCancellationRequested();
+                    Stream contentStream = state.RequestContentStream;
+
+                    // TODO: Handle chunked-mode
+
+                    int bytesRead = 0;
+                    while (bytesRead < (int)totalSize)
+                    {
+                        state.CancellationToken.ThrowIfCancellationRequested();
+                        var byteBuffer = new byte[totalSize];
+                        int numBytes = contentStream.Read(byteBuffer, bytesRead, (int)totalSize - bytesRead);
+                        if (numBytes == 0)
+                        {
+                            break;
+                        }
+                        Marshal.Copy(byteBuffer, bytesRead, new IntPtr(ptr.ToInt64() + bytesRead), numBytes);
+                        bytesRead += numBytes;
+                    }
+#if DEBUG
+                    LogCurlBuffer(ptr, (ulong)bytesRead);
+#endif
+                    return (ulong)bytesRead;
+                });
+        }
+
+        private int CurlSendIoCtlCallback(
+            IntPtr handle,
+            int cmd,
+            IntPtr context)
+        {
+            if (cmd != curliocmd.CURLIOCMD_RESTARTREAD)
+            {
+                return curlioerr.CURLIOE_UNKNOWNCMD;
+            }
+
+            GCHandle gch = GCHandle.FromIntPtr(context);
+            var state = (RequestCompletionSource)gch.Target;
+
+            try
+            {
+                CheckDisposed();
+                state.CancellationToken.ThrowIfCancellationRequested();
+
+                Stream contentStream = state.RequestContentStream;
+                if (!contentStream.CanSeek)
+                {
+                    return curlioerr.CURLIOE_FAILRESTART;
+                }
+                contentStream.Seek(0, SeekOrigin.Begin);
+                return curlioerr.CURLIOE_OK;
+            }
+            catch (Exception ex)
+            {
+                state.SavedException = ex;
+#if DEBUG
+                if (Logging.On) Logging.Exception(Logging.Http, this, "CurlSendIoCtlCallback exception for " + Logging.GetObjectLogHash(state.ResponseMessage.ContentStream) + ": ", ex);
+#endif
+                return curlioerr.CURLIOE_FAILRESTART; // Force curl_easy_perform to fail
+            }
```

Comment:
It gets freed in finally block of StartRequest. Is a free required everytime FromIntPtr is used?